File: src/test/java/no/fintlabs/flyt/gateway/application/archive/dispatch/web/FintArchiveDispatchClientTest.java
package no.fintlabs.flyt.gateway.application.archive.dispatch.web;

import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
import no.fintlabs.flyt.gateway.application.archive.WebUtilErrorHandler;
import no.fintlabs.flyt.gateway.application.archive.resource.web.FintArchiveResourceClient;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.http.ResponseEntity;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientException;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import java.net.URI;
import java.time.Duration;
import java.util.concurrent.atomic.AtomicLong;

import static org.mockito.Mockito.*;
import static org.springframework.web.reactive.function.client.WebClient.RequestHeadersUriSpec;

class FintArchiveDispatchClientTest {

    WebClient fintWebClient;
    FintArchiveResourceClient fintArchiveResourceClient;
    FintArchiveDispatchClient fintArchiveDispatchClient;
    MeterRegistry meterRegistry;
    WebUtilErrorHandler webUtilErrorHandler;

    @BeforeEach
    public void setup() {
        fintWebClient = mock(WebClient.class);
        fintArchiveResourceClient = mock(FintArchiveResourceClient.class);
        meterRegistry = new SimpleMeterRegistry();
        webUtilErrorHandler = mock(WebUtilErrorHandler.class);
        fintArchiveDispatchClient = new FintArchiveDispatchClient(
                FintArchiveDispatchClientConfigurationProperties
                        .builder()
                        .postFileTimeoutMillis(130000L)
                        .postRecordTimeoutMillis(130000L)
                        .postCaseTimeoutMillis(130000L)
                        .getStatusTimeoutMillis(130000L)
                        .createdLocationPollBackoffMinDelayMillis(100L)
                        .createdLocationPollBackoffMaxDelayMillis(250L)
                        .createdLocationPollTotalTimeoutMillis(1000L)
                        .build(),
                fintWebClient,
                fintArchiveResourceClient,
                meterRegistry,
                webUtilErrorHandler
        );
    }

    @Test
    public void givenCreatedLocationOnFirstPoll_whenPollForCreatedLocation_shouldReturnWithoutRepeats() {
        WebClient.RequestHeadersUriSpec requestHeadersUriSpec = mock(RequestHeadersUriSpec.class);
        when(fintWebClient.get()).thenReturn(requestHeadersUriSpec);

        WebClient.RequestBodySpec requestBodySpec = mock(WebClient.RequestBodySpec.class);
        when(requestHeadersUriSpec.uri(URI.create("testStatusUri"))).thenReturn(requestBodySpec);

        when(requestBodySpec.httpRequest(any())).thenReturn(requestBodySpec);

        WebClient.ResponseSpec responseSpec = mock(WebClient.ResponseSpec.class);
        when(requestBodySpec.retrieve()).thenReturn(responseSpec);

        when(responseSpec.toBodilessEntity()).thenAnswer(invocation -> Mono.just(ResponseEntity.created(URI.create("testCreatedUri")).build()));

        StepVerifier.create(fintArchiveDispatchClient.pollForCreatedLocation(URI.create("testStatusUri")))
                .expectNext(URI.create("testCreatedUri"))
                .expectComplete()
                .verify();

        verify(fintWebClient, times(1)).get();
    }

    @Test
    public void givenRequiredRepeatsFasterThanMaxTotalTimeout_whenPollForCreatedLocation_shouldPollForStatusLocationUntilProvided() {
        WebClient.RequestHeadersUriSpec requestHeadersUriSpec = mock(RequestHeadersUriSpec.class);
        when(fintWebClient.get()).thenReturn(requestHeadersUriSpec);

        WebClient.RequestBodySpec requestBodySpec = mock(WebClient.RequestBodySpec.class);
        when(requestHeadersUriSpec.uri(URI.create("testStatusUri"))).thenReturn(requestBodySpec);

        when(requestBodySpec.httpRequest(any())).thenReturn(requestBodySpec);

        WebClient.ResponseSpec responseSpec = mock(WebClient.ResponseSpec.class);
        when(requestBodySpec.retrieve()).thenReturn(responseSpec);

        AtomicLong startTime = new AtomicLong(0);
        when(responseSpec.toBodilessEntity()).thenAnswer(invocation -> {
            if (startTime.get() == 0) {
                startTime.set(System.nanoTime());
                return Mono.just(ResponseEntity.noContent().build());
            }
            if (Duration.ofNanos(System.nanoTime()).minus(Duration.ofNanos(startTime.get()))
                        .compareTo(Duration.ofMillis(500)) < 0) {
                return Mono.just(ResponseEntity.noContent().build());
            }
            return Mono.just(ResponseEntity.created(URI.create("testCreatedUri")).build());
        });

        StepVerifier.create(fintArchiveDispatchClient.pollForCreatedLocation(URI.create("testStatusUri")))
                .expectNext(URI.create("testCreatedUri"))
                .expectComplete()
                .verify();

        verify(fintWebClient, atLeast(2)).get();
    }

    @Test
    public void givenMaxPollingTimeIsExceeded_whenPollForCreatedLocation_shouldThrowException() {
        WebClient.RequestHeadersUriSpec requestHeadersUriSpec = mock(RequestHeadersUriSpec.class);
        when(fintWebClient.get()).thenReturn(requestHeadersUriSpec);

        WebClient.RequestBodySpec requestBodySpec = mock(WebClient.RequestBodySpec.class);
        when(requestHeadersUriSpec.uri(URI.create("testStatusUri"))).thenReturn(requestBodySpec);

        when(requestBodySpec.httpRequest(any())).thenReturn(requestBodySpec);

        WebClient.ResponseSpec responseSpec = mock(WebClient.ResponseSpec.class);
        when(requestBodySpec.retrieve()).thenReturn(responseSpec);

        when(responseSpec.toBodilessEntity()).thenAnswer(invocation -> Mono.just(ResponseEntity.noContent().build()));

        StepVerifier.create(fintArchiveDispatchClient.pollForCreatedLocation(URI.create("testStatusUri")))
                .expectErrorMessage("Reached max total timeout for polling created location from destination")
                .verify();

        verify(fintWebClient, atLeast(2)).get();
    }

    @Test
    public void givenErrorResponse_whenPollForCreatedLocation_shouldThrowException() {
        WebClient.RequestHeadersUriSpec requestHeadersUriSpec = mock(RequestHeadersUriSpec.class);
        when(fintWebClient.get()).thenReturn(requestHeadersUriSpec);

        WebClient.RequestBodySpec requestBodySpec = mock(WebClient.RequestBodySpec.class);
        when(requestHeadersUriSpec.uri(URI.create("testStatusUri"))).thenReturn(requestBodySpec);

        when(requestBodySpec.httpRequest(any())).thenReturn(requestBodySpec);

        WebClient.ResponseSpec responseSpec = mock(WebClient.ResponseSpec.class);
        when(requestBodySpec.retrieve()).thenReturn(responseSpec);

        WebClientException webClientException = mock(WebClientException.class);
        when(webClientException.getMessage()).thenReturn("test message");
        when(responseSpec.toBodilessEntity()).thenAnswer(invocation -> Mono.error(webClientException));

        StepVerifier.create(fintArchiveDispatchClient.pollForCreatedLocation(URI.create("testStatusUri")))
                .expectErrorMessage("test message")
                .verify();

        verify(fintWebClient, times(1)).get();
    }

}

File: src/test/kotlin/no/novari/flyt/archive/gateway/dispatch/web/FintArchiveDispatchClientTest.kt
package no.novari.flyt.archive.gateway.dispatch.web

import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.simple.SimpleMeterRegistry
import no.novari.flyt.archive.gateway.WebUtilErrorHandler
import no.novari.flyt.archive.gateway.resource.web.FintArchiveResourceClient
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.atLeast
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.doThrow
import org.mockito.kotlin.mock
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.springframework.http.ResponseEntity
import org.springframework.web.client.ResourceAccessException
import org.springframework.web.client.RestClient
import java.net.URI
import java.time.Duration

class FintArchiveDispatchClientTest {
    private lateinit var fintRestClient: RestClient
    private lateinit var fintArchiveResourceClient: FintArchiveResourceClient
    private lateinit var fintArchiveDispatchClient: TestableFintArchiveDispatchClient
    private lateinit var meterRegistry: MeterRegistry
    private lateinit var webUtilErrorHandler: WebUtilErrorHandler

    @BeforeEach
    fun setup() {
        fintRestClient = mock()
        fintArchiveResourceClient = mock()
        meterRegistry = SimpleMeterRegistry()
        webUtilErrorHandler = mock()
        val properties =
            FintArchiveDispatchClientConfigurationProperties().apply {
                createdLocationPollBackoffMinDelay = Duration.ofMillis(100L)
                createdLocationPollBackoffMaxDelay = Duration.ofMillis(250L)
                createdLocationPollTotalTimeout = Duration.ofMillis(1000L)
            }
        fintArchiveDispatchClient =
            TestableFintArchiveDispatchClient(
                properties,
                fintRestClient,
                fintArchiveResourceClient,
                meterRegistry,
                webUtilErrorHandler,
            )
    }

    @Test
    fun `given a created location on the first poll, returns without repeating`() {
        mockGetReturning(ResponseEntity.created(URI.create("testCreatedUri")).build())

        val result = fintArchiveDispatchClient.pollForCreatedLocationPublic(URI.create("testStatusUri"))

        assertThat(result).isEqualTo(URI.create("testCreatedUri"))
        verify(fintRestClient, times(1)).get()
    }

    @Test
    fun `given the max polling time is exceeded, throws CreatedLocationPollTimeoutException`() {
        mockGetReturning(ResponseEntity.noContent().build())

        assertThatThrownBy { fintArchiveDispatchClient.pollForCreatedLocationPublic(URI.create("testStatusUri")) }
            .isInstanceOf(CreatedLocationPollTimeoutException::class.java)
            .hasMessageContaining("statusUri=testStatusUri")
            .hasMessageContaining("lastStatus=204 NO_CONTENT")

        verify(fintRestClient, atLeast(2)).get()
    }

    @Test
    fun `given an error response, propagates the exception`() {
        mockGetThrowing(ResourceAccessException("test message"))

        assertThatThrownBy { fintArchiveDispatchClient.pollForCreatedLocationPublic(URI.create("testStatusUri")) }
            .isInstanceOf(ResourceAccessException::class.java)
            .hasMessage("test message")
    }

    private fun mockGetReturning(response: ResponseEntity<Void>) {
        val uriSpec = mock<RestClient.RequestHeadersUriSpec<*>>()
        val headersSpec = mock<RestClient.RequestHeadersSpec<*>>()
        val responseSpec = mock<RestClient.ResponseSpec>()
        whenever(fintRestClient.get()).thenReturn(uriSpec)
        doReturn(headersSpec).whenever(uriSpec).uri(any<URI>())
        whenever(headersSpec.retrieve()).thenReturn(responseSpec)
        whenever(responseSpec.toBodilessEntity()).thenReturn(response)
    }

    private fun mockGetThrowing(error: Throwable) {
        val uriSpec = mock<RestClient.RequestHeadersUriSpec<*>>()
        val headersSpec = mock<RestClient.RequestHeadersSpec<*>>()
        val responseSpec = mock<RestClient.ResponseSpec>()
        whenever(fintRestClient.get()).thenReturn(uriSpec)
        doReturn(headersSpec).whenever(uriSpec).uri(any<URI>())
        whenever(headersSpec.retrieve()).thenReturn(responseSpec)
        doThrow(error).whenever(responseSpec).toBodilessEntity()
    }

    private class TestableFintArchiveDispatchClient(
        properties: FintArchiveDispatchClientConfigurationProperties,
        fintRestClient: RestClient,
        fintArchiveResourceClient: FintArchiveResourceClient,
        meterRegistry: MeterRegistry,
        webUtilErrorHandler: WebUtilErrorHandler,
    ) : FintArchiveDispatchClient(
            properties,
            fintRestClient,
            fintArchiveResourceClient,
            meterRegistry,
            webUtilErrorHandler,
        ) {
        fun pollForCreatedLocationPublic(statusUri: URI) = pollForCreatedLocation(statusUri)
    }
}
